post-processor/vagrant-cloud/post-processor.go

```diff
@@ -150,12 +153,21 @@ func (p *PostProcessor) PostProcess(ctx context.Context, ui packersdk.Ui, artifa
 		ui.Message("Warning: Using Vagrant Cloud token found in ATLAS_TOKEN. Please make sure it is correct, or set VAGRANT_CLOUD_TOKEN")
 	}
 
+	// Used for caching metadata extracted from artifact if required
+	var metadata map[string]interface{}
+
 	// Determine the name of the provider for Vagrant Cloud, and Vagrant
-	providerName, err := getProvider(artifact.Id(), artifact.Files()[0], builtins[artifact.BuilderId()])
+	providerName, err := getProvider(artifact.Id(), artifact.Files()[0], builtins[artifact.BuilderId()], &metadata)
```

Comment:
I see what you are doing here with the metadata map and why you have the nil checks in other functions. Since you are technically reading this data once and using a guard to avoid re-reading the file I think we can change this up a bit to avoid passing around the metadata map for marshaling and reduce a number of nil checks. This of course breaks the testing but that should be easy to fix. Let me know what you think . 

For this suggestion I'm going to change the signature of the 
```go

func (p *PostProcessor) PostProcess(ctx context.Context, ui packersdk.Ui, artifact packersdk.Artifact) (packersdk.Artifact, bool, bool, error) {
 
// Redacted code above to highlight changes ....

	// Used for caching metadata extracted from artifact if required
	metadata, err := metadataFromVagrantBox(artifact.Files()[0])
	if err != nil {
		return nil, false, false, err
	}

	// Determine the name of the provider for Vagrant Cloud, and Vagrant
	providerName, err := getProvider(artifact.Id(), builtins[artifact.BuilderId()], metadata)
	if err != nil {
		return nil, false, false, fmt.Errorf("error getting provider name: %s", err)
	}

	// Determine the name of the architecture for Vagrant Cloud, and Vagrant
	archName, err := getArchitecture(p.config, metadata)
	if err != nil {
		return nil, false, false, fmt.Errorf("error getting architecture name: %w", err)
	}
 
// Redacted the lines that follow to just so the suggested changes. 

	return NewArtifact(providerName, p.config.Tag), true, false, nil
}

// getArchitecture was merged with architectureFromVagrantBox
func getArchitecture(config Config, metadata map[string]interface{}) (string, error) {
	// If the architecture was defined in the config, use that value
	if config.Architecture != "" {
		return config.Architecture, nil
	}

	// Check for the architecture within the artifact metadata
	if arch, ok := metadata["architecture"]; ok {
		if architectureName, ok := arch.(string); ok && architectureName != "" {
			return architectureName, nil
		}
	}
	return "", fmt.Errorf("Error: Could not determine architecture from box metadata.json file")
}

// getProvider was merged with providerFromVagrantBox.
func getProvider(builderName, builderId string, metadata map[string]interface{}) (string, error) {
	if builderId != "artifice" {
		// For the Vagrant builder and Vagrant post processor the provider can
		// be determined from information embedded in the artifact
		return providerFromBuilderName(builderName), nil
	}

	// The artifice post processor cannot embed any data in the
	// supplied artifact so the provider information must be extracted
	// from the box file directly
	if prov, ok := metadata["provider"]; ok {
		if providerName, ok := prov.(string); ok && providerName != "" {
			return providerName, nil
		}
	}
	return "", fmt.Errorf("Error: Could not determine provider from box metadata.json file")
}

// Returns the metadata found within the metadata file
// packaged inside the box
func metadataFromVagrantBox(boxfile string) (map[string]interface{}, error) {
	log.Printf("Attempting to extract metadata in box file. This may take some time...")

	f, err := os.Open(boxfile)
	if err != nil {
		return nil, fmt.Errorf("Error attempting to open box file: %w", err)
	}
	defer f.Close()

	// Vagrant boxes are gzipped tar archives
	ar, err := gzip.NewReader(f)
	if err != nil {
		return nil, fmt.Errorf("Error unpacking box archive: %w", err)
	}
	tr := tar.NewReader(ar)

	metadata := make(map[string]interface{})
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			return nil, fmt.Errorf("Error: metadata.json file not found in box: %s", boxfile)
		}

		if err != nil {
			return nil, fmt.Errorf("Error reading header info from box tar archive: %w", err)
		}

		if hdr.Name == "metadata.json" {
			contents, err := ioutil.ReadAll(tr)
			if err != nil {
				return nil, fmt.Errorf("Error reading contents of metadata.json file from box file: %w", err)
			}
			err = json.Unmarshal(contents, metadata)
			if err != nil {
				return nil, fmt.Errorf("Error parsing metadata.json file: %w", err)
			}

			break
		}
	}

	return metadata, nil
}
```